pilot/pkg/model/xds_cache.go

```diff
@@ -164,20 +188,67 @@ func newLru(evictCallback simplelru.EvictCallback) simplelru.LRUCache {
 	return l
 }
 
-func (l *lruCache) evict(k interface{}, v interface{}) {
+// This is the callback passed to LRU, it will be called whenever a key is removed.
+func (l *LruCache) onEvict(k interface{}, _ interface{}) {
 	if features.EnableXDSCacheMetrics {
 		xdsCacheEvictions.Increment()
 	}
 
-	key := k.(string)
-	value := v.(cacheValue)
+	if l.trackEvicted {
+		l.evictedKeys.Insert(k.(string))
+		select {
+		case l.evictCh <- struct{}{}:
+		default: // the signal has not been consumed
+		}
+	}
+}
+
+func (l *LruCache) handleEvicted(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-l.evictCh:
+			l.clearEvicted()
+		case <-stopCh:
+			log.Infof("LruCache has been stopped")
+			return
+		}
+	}
+}
+
+// clearEvicted is to clear configIndex and typesIndex based on evictedKeys.gi
+func (l *LruCache) clearEvicted() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	// In order to save cpu, if the evicted keys is very small, we can just ignore
+	// TODO: make it configurable
+	if len(l.evictedKeys) < 100 {
```

Comment:
is the optimization worth the inaccuracy? I think things like delta will depend on whether or not entries in the cache are still valid. We could just check the cache AND evictedKeys though -- @howardjohn

Reply:
Not sure i understand

Reply:
This can be seen as a debouncer, hope you can better understand